pkgs/tools/typesetting/tex/texlive/build-tex-env.sh

```diff
@@ -0,0 +1,327 @@
+        read -r type map <<<"$line"
+        # if a line for $map exists, remove it to avoid
+        # pointless warnings from updmap
+        sedExpr="$sedExpr /^(#! )?[^ ]*Map $map$/d;"
+    done <<<"$updmapCfg"
+
+    # disable all the remaining font maps
+    sedExpr="$sedExpr /^[^ ]*Map/{ s/^/#! /p };"
+
+    {
+        echo "# Generated by nixpkgs"
+        sed -E -n -e "$sedExpr" "$TEXMFDIST"/web2c/updmap.cfg
+        [[ -z $updmapCfg ]] || printf '%s' "$updmapCfg"
+    } >"$outFile"
+}
+
+_create_config_files () {
+    # simplified arguments
+    local header="$1"
+    local dest="$2"
+    local prefix="$3"
+    local lines="$4"
+    local suffix="$5"
+    if [[ -z "$header" || -e "$header" ]] ; then
+        nixpkgs_texlive_log "writing ${dest##*/} to $dest"
+        {
+            [[ -z $prefix ]] || printf '%s\n' "$prefix"
+            [[ ! -e $header ]] || cat "$header"
+            [[ -z $lines ]] || printf '%s\n' "$lines"
+            [[ -z $suffix ]] || printf '%s\n' "$suffix"
+        } >"$dest"
+    fi
+}
+
+create_language_dat () {
+    _create_config_files \
+        "$TEXMFDIST"/tex/generic/config/language.us \
+        "$TEXMFSYSVAR"/tex/generic/config/language.dat \
+        '% Generated by nixpkgs' \
+        "$languageDat" \
+        ''
+}
+
+create_language_def () {
+    _create_config_files \
+        "$TEXMFDIST"/tex/generic/config/language.us.def \
+        "$TEXMFSYSVAR"/tex/generic/config/language.def \
+        '' \
+        "$languageDef" \
+        '%%% No changes may be made beyond this point.
+
+\uselanguage {USenglish}             %%% This MUST be the last line of the file.'
+}
+
+create_language_lua () {
+    _create_config_files \
+        "$TEXMFDIST"/tex/generic/config/language.us.lua \
+        "$TEXMFSYSVAR"/tex/generic/config/language.dat.lua \
+        '-- Generated by nixpkgs' \
+        "$languageLua" \
+        '}'
+}
+
+update_context_cache () {
+    if [[ -x $out/bin/mtxrun ]] ; then
+        nixpkgs_texlive_log "generating ConTeXt cache"
+
+        # temporarily patch mtxrun.lua to generate uuid's deterministically from SOURCE_DATE_EPOCH
+        mv "$out"/bin/mtxrun.lua{,.orig}
+        substitute "$out"/bin/mtxrun.lua.orig "$out"/bin/mtxrun.lua \
+            --replace-fail 'randomseed(math.initialseed)' "randomseed($SOURCE_DATE_EPOCH)"
+
+        # this is very verbose, save the output for debugging purposes
+        {
+            mtxrun --generate
+            context --luatex --generate
+            [[ ! -x $out/bin/luajittex ]] || context --luajittex --generate
+        } >"$TEXMFSYSVAR"/context-cache.log
+
+        mv "$out"/bin/mtxrun.lua{.orig,}
+    fi
+}
+
+# init variables (export only the ones that are used in the wrappers)
+TEXMFROOT="$texmfroot"
+TEXMFDIST="$texmfdist"
+export PATH="$out/bin:$PATH"
+TEXMFSYSCONFIG="$out/share/texmf-config"
+TEXMFSYSVAR="$out/share/texmf-var"
+export TEXMFCNF="$TEXMFSYSCONFIG/web2c"
+
+do_postinst_stuff
```

Comment:
Perhaps convert this to a shellHook `postBuildHooks+=(do_postinst_stuff)` (possibly give it a more texlive-related name)

Reply:
> Perhaps convert this to a shellHook `postBuildHooks+=(do_postinst_stuff)` (possibly give it a more texlive-related name)

Ah, that's not working... `buildEnv` is based on `runCommand`, which sets `buildCommand`, which bypasses phases and hooks altogether.

I can prefix all function names with `tl_`, though, for future proofing?

Reply:
(To clarify: `do_postinst_stuff` is literally the name of the corresponding function in `install-tl`, I didn't make it up.)

Reply:
This made me think that we could use `mkDerivation` directly, and group those functions under traditional phases (configure for `fmtutil.cnf`, build for the formats, install or fixup for the wrappers). We could then switch to the `finalAttrs: prevAttrs: ...` pattern and drop `.overrideTeXConfig`.

Something for 24.11, maybe. I'd rather make sure the current refactor really works, do other minor things (I have tlcontrib ready!), then start experimenting with the 'overrideable texlive' idea right after 24.05.